shotover-proxy/src/frame/kafka.rs

```diff
@@ -0,0 +1,129 @@
+use crate::codec::kafka::RequestHeader as CodecRequestHeader;
+use anyhow::{anyhow, Context, Result};
+use bytes::{BufMut, Bytes, BytesMut};
+use kafka_protocol::messages::{
+    ApiKey, ProduceRequest, ProduceResponse, RequestHeader, ResponseHeader,
+};
+use kafka_protocol::protocol::{Decodable, Encodable};
+
+// No way to know which version to use, just have to guess
+const REQUEST_HEADER_VERSION: i16 = 2;
```

Comment:
_Kafka has a "bidirectional" client compatibility policy. In other words, new clients can talk to old servers, and old clients can talk to new servers. This allows users to upgrade either clients or servers without experiencing any downtime.

Since the Kafka protocol has changed over time, clients and servers need to agree on the schema of the message that they are sending over the wire. This is done through API versioning.

Before each request is sent, the client sends the API key and the API version. These two 16-bit numbers, when taken together, uniquely identify the schema of the message to follow.

The intention is that clients will support a range of API versions. When communicating with a particular broker, a given client should use the highest API version supported by both and indicate this version in their requests.

The server will reject requests with a version it does not support, and will always respond to the client with exactly the protocol format it expects based on the version it included in its request. The intended upgrade path is that new features would first be rolled out on the server (with the older clients not making use of them) and then as newer clients are deployed these new features would gradually be taken advantage of._

So it looks like we pick the highest we support, and then have to negotiate down if the server comes back a different protocol format (e.g. in the form of the two 16 bit numbers)

Reply:
From https://kafka.apache.org/protocol.html#protocol_compatibility

Reply:
Also supports and ApiVersionsRequest - https://kafka.apache.org/protocol.html#api_versions

Reply:
> So it looks like we pick the highest we support, and then have to negotiate down if the server comes back a different protocol format (e.g. in the form of the two 16 bit numbers)


That is correct and we already handle that correctly, or at least we dont get in the way of the client/server handling that correctly.
What my code comment is talking about is the version used for the header which doesnt seem to have any negotiation in the protocol. The driver I was referring to `rskafka` just hardcoded it to the versions I am hardcoding here.